cv/lane_detection/src/lane_detection_inferencer.cpp

```diff
@@ -0,0 +1,242 @@
+constexpr int NETWORK_OUTPUT_HEIGHT{180};
+
+template <typename T> T vectorProduct(const std::vector<T> &v)
+{
+    return std::accumulate(v.begin(), v.end(), 1, std::multiplies<T>());
+}
+
+uint8_t getMedian(std::vector<uint8_t> input)
+{
+    if (input.empty()) {
+        throw std::invalid_argument("Received empty input vector!");
+    }
+
+    std::nth_element(input.begin(), input.begin() + input.size() / 2,
+                     input.end());
+    return input[input.size() / 2];
+}
+
+std::vector<uint8_t> convertToQuadrant(cv::Mat input, const int r_min,
+                                       const int r_max, const int c_min,
+                                       const int c_max)
+{
+    std::vector<uint8_t> output;
+    for (int r = r_min; r < r_max; ++r) {
+        for (int c = c_min; c < c_max; ++c) {
+            output.emplace_back(input.at<uchar>(r, c));
+        }
+    }
+    return output;
+}
+
+cv::Mat getQuadrantEdges(cv::Mat grayscaled_image, const int row_start,
+                         const int row_end, const int col_start,
+                         const int col_end)
+{
+    const std::vector<uint8_t> quad{convertToQuadrant(
+        grayscaled_image, row_start, row_end, col_start, col_end)};
+
+    const double median{static_cast<double>(getMedian(quad))};
+    const double lower_bound{std::max(0.0, std::floor((1 - 0.205) * median))};
+    const double upper_bound{std::min(255.0, std::floor((1 + 0.205) * median))};
+    cv::Mat quad_cv_mat{
+        cv::Mat(quad, CV_8UC1)
+            .reshape(1, static_cast<int>(grayscaled_image.rows / 2))};
+    cv::Mat edges;
+    cv::Canny(quad_cv_mat, edges, lower_bound, upper_bound);
+    return edges;
+}
+
+cv::Mat getEdgeChannel(cv::Mat grayscaled_image)
+{
+    cv::Mat quad1_edges{getQuadrantEdges(
+        grayscaled_image, 0, static_cast<int>(grayscaled_image.rows / 2), 0,
+        static_cast<int>(grayscaled_image.cols / 2))};
+    cv::Mat quad2_edges{getQuadrantEdges(
+        grayscaled_image, 0, static_cast<int>(grayscaled_image.rows / 2),
+        static_cast<int>(grayscaled_image.cols / 2), grayscaled_image.cols)};
+    cv::Mat quad3_edges{getQuadrantEdges(
+        grayscaled_image, static_cast<int>(grayscaled_image.rows / 2),
+        grayscaled_image.rows, static_cast<int>(grayscaled_image.cols / 2),
+        grayscaled_image.cols)};
+    cv::Mat quad4_edges{getQuadrantEdges(
+        grayscaled_image, static_cast<int>(grayscaled_image.rows / 2),
+        grayscaled_image.rows, 0, static_cast<int>(grayscaled_image.cols / 2))};
+
+    // stitch edges together
+    const int new_rows{std::max(quad1_edges.rows + quad3_edges.rows,
+                                quad2_edges.rows + quad4_edges.rows)};
+    const int new_cols{std::max(quad1_edges.cols + quad3_edges.cols,
+                                quad2_edges.cols + quad4_edges.cols)};
+
+    cv::Mat edges_mask = cv::Mat::zeros(new_rows, new_cols, CV_8UC1);
+
+    quad1_edges.copyTo(
```

Comment:
Copies here might not be necessary @spencerteetaert 

I didn't bother trying to optimize but do you think we should try to avoid it ?

Reply:
there definitely is room for optimization. I'm not too too concerned with it already running at 40+fps. We can spend more time on it if it's under performing with full system running

Reply:
Agreed